### PR TITLE
Serialize query AST to JSON.

### PR DESCRIPTION
**Pull request by jeschkies**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Special notes for your reviewer**:

**Checklist**
- [ ] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [ ] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/logql/syntax/walk.go [hunk 1]
```diff
@@ -11,3 +13,112 @@ func walkAll(f WalkFn, xs ...Walkable) {
...
+	VisitLabelFilter(*LabelFilterExpr)
+	VisitLabelFmt(*LabelFmtExpr)
+	VisitLabelParser(*LabelParserExpr)
+	VisitLineFilter(*LineFilterExpr)
+	VisitLineFmt(*LineFmtExpr)
+	VisitLogfmtExpressionParser(*LogfmtExpressionParser)
+	VisitLogfmtParser(*LogfmtParserExpr)
+}
+
+func Dispatch(root Expr, v RootVisitor) error {
```
Comment (jeschkies) [inline pkg/logql/syntax/walk.go:53]: This follows https://www.lihaoyi.com/post/ZeroOverheadTreeProcessingwiththeVisitorPattern.html. We could also add `Accept(Visitor)` to all expression like the `Prettify` method. This would avoid pattern matches.

File: pkg/logql/serialize_test.go [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+			require.NoError(t, err)
+
+			var buf bytes.Buffer
+			err = EncodeJSON(expr, &buf)
+			require.NoError(t, err)
+
+			actual, err := DecodeJSON(buf.String())
+			require.NoError(t, err)
+
+			fmt.Println(buf.String())
```
Comment (dannykopping) [inline pkg/logql/serialize_test.go:44]:
```suggestion
```

File: pkg/logql/syntax/ast.go [hunk 3]
```diff
@@ -53,11 +53,15 @@ type implicit struct{}
...
+//
+//sumtype:decl
 type LogSelectorExpr interface {
 	Matchers() []*labels.Matcher
-	LogPipelineExpr
+	Pipeline() (Pipeline, error)
 	HasFilter() bool
 	Expr
+
+	isLogSelectorExpr()
```
Comment (dannykopping) [inline pkg/logql/syntax/ast.go:64]: Is this still required? I don't see it in use anywhere

File: pkg/logql/syntax/ast.go [hunk 4]
```diff
@@ -66,19 +70,17 @@ type (
...
-}
-
 // StageExpr is an expression defining a single step into a log pipeline
+//
+//sumtype:decl
 type StageExpr interface {
 	Stage() (log.Stage, error)
 	Expr
+
+	isStageExpr()
```
Comment (dannykopping) [inline pkg/logql/syntax/ast.go:80]: Ditto, is this used?

**Review by dannykopping — COMMENTED** [review]
Looks great! I think we'll need to flesh out the tests a bit more before we land this though.

I also see changes to `vendor` without related changes to `go.*`?

File: pkg/logql/syntax/ast.go (hunk 3 above)
Comment (jeschkies) [reply pkg/logql/syntax/ast.go:64]: This seals the interface and makes it easier to determine which expression implements it. The Protobuf generation is doing something similar for `oneof` and the expressions themselves in our syntax packages does so as well.

**Comment by jeschkies** [comment]
>  I think we'll need to flesh out the tests a bit more before we land this though.

For sure. The idea would be to check in some sanitized queries we see.

File: pkg/logql/syntax/parser_test.go [hunk 5]
```diff
@@ -17,241 +17,240 @@ func NewStringLabelFilter(s string) *string {
 	return &s
 }
 
-func TestParse(t *testing.T) {
```
Comment (jeschkies) [inline pkg/logql/syntax/parser_test.go:20]: This is the biggest change so that I can reuse the test cases. Should I split this out into another PR?

File: pkg/logql/syntax/serialize_test.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+import (
+	"bytes"
+	"strings"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestJSONSerializationRoundTrip(t *testing.T) {
```
Comment (dannykopping) [inline pkg/logql/syntax/serialize_test.go:11]: Do we have a reasonable amount of coverage with these tests?

**Review by dannykopping — APPROVED** [review]
This is looking **really** slick!
Added a couple nits, great work @jeschkies

File: pkg/logql/syntax/parser_test.go (hunk 5 above)
Comment (jeschkies) [reply pkg/logql/syntax/parser_test.go:20]: With biggest I've meant the most line changes because the indention changed. That's about it.

File: pkg/logql/syntax/serialize_test.go (hunk 6 above)
Comment (jeschkies) [reply pkg/logql/syntax/serialize_test.go:11]: These are 153 test cases
```
> go test -v -run TestJSONSerializationParseTestCases ./pkg/logql/syntax/... | rg PASS | sort | uniq | wc -l
152
```

they are *the* cases for the parser. That said there might be some edge cases. I was planning to take some actual queries, sanitize them and the add them. If it's ok I'll do this afterwards as I need the visitor for the sanitizing.

File: pkg/logql/syntax/parser_test.go (hunk 5 above)
Comment (dannykopping) [reply pkg/logql/syntax/parser_test.go:20]: Ah, right. I review without whitespace, so that explains the confusion

File: pkg/logql/log/ip.go [hunk 7]
```diff
@@ -78,39 +78,41 @@ func (f *IPLineFilter) filterTy(line []byte, ty labels.MatchType) bool {
 
 type IPLabelFilter struct {
 	ip *ipFilter
-	ty LabelFilterType
+	Ty LabelFilterType
```
Comment (cstyan) [inline pkg/logql/log/ip.go:81]: does uppercase `Type` conflict with a built in or something?

File: pkg/logql/log/label_filter.go [hunk 8]
```diff
@@ -54,23 +54,28 @@ func (f LabelFilterType) String() string {
...
 
 // LabelFilterer can filter extracted labels.
+//
+//sumtype:decl
 type LabelFilterer interface {
 	Stage
 	fmt.Stringer
+
+	// Seal trait
+	isLabelFilterer()
```
Comment (cstyan) [inline pkg/logql/log/label_filter.go:64]: what are the comments here?

File: pkg/logql/syntax/ast.go (hunk 3 above)
Comment (cstyan) [reply pkg/logql/syntax/ast.go:65]: can we include a comment here that explains this? or is there something I'm missing somewhere else that makes it more obvious

File: pkg/logql/syntax/serialize.go [hunk 9]
```diff
@@ -0,0 +1,878 @@
...
+		return decodeVectorAgg(iter)
+	case "range_agg":
+		return decodeRangeAgg(iter)
+	case "literal":
+		return decodeLiteral(iter)
+	case "vector":
+		return decodeVector(iter)
+	case "label_replace":
+		return decodeLabelReplace(iter)
+	case "log_selector":
```
[the remark this answers is not in the document]
Comment (cstyan) [reply pkg/logql/syntax/serialize.go:49]: +1

File: pkg/logql/syntax/serialize.go [hunk 10]
```diff
@@ -0,0 +1,878 @@
...
+func (v *JSONSerializer) VisitBinOp(e *BinOpExpr) {
+	v.WriteObjectStart()
+
+	v.WriteObjectField("bin")
+	v.WriteObjectStart()
+
+	v.WriteObjectField("op")
+	v.WriteString(e.Op)
+
+	v.WriteMore()
```
Comment (cstyan) [inline pkg/logql/syntax/serialize.go:67]: what does `WriteMore` do?

File: pkg/logql/syntax/walk.go (hunk 1 above)
Comment (cstyan) [reply pkg/logql/syntax/walk.go:59]: can you elaborate? I'm not sure what this means

File: pkg/logql/syntax/serialize.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/logql/syntax/serialize.go:49]: I've missed this comment. We'll follow up.

[the remark this answers is not in the document]
Comment (jeschkies) [reply pkg/logql/syntax/serialize.go:49]: Thanks. I actually did find two errors https://github.com/grafana/loki/pull/11230

File: pkg/logql/syntax/serialize_test.go (hunk 6 above)
Comment (jeschkies) [reply pkg/logql/syntax/serialize_test.go:11]: Thanks. I actually did find two errors https://github.com/grafana/loki/pull/11230
